lib/mllp/client.ex

```diff
@@ -0,0 +1,444 @@
+          socket_opts: map()
+        ]
+
+  @type send_options :: %{
+          optional(:expect_reply) => boolean(),
+          optional(:timeout) => non_neg_integer(),
+          optional(:wrap_message) => boolean(),
+          optional(:reply_timeout) => non_neg_integer() | :infinity,
+          optional(:verify_reply) => boolean()
+        }
+
+  @callback send(
+              pid :: pid,
+              payload :: HL7.Message.t() | String.t(),
+              options :: send_options()
+            ) ::
+              {:ok, String.t()}
+              | MLLP.Ack.ack_verification_result()
+              | {:error, sender_error()}
+end
+
+defmodule MLLP.Client do
+  use GenServer
+  require Logger
+
+  alias MLLP.{Envelope, Ack, ClientContract, TCP, TLS}
+
+  @behaviour ClientContract
+
+  @type pid_ref :: atom | pid | {atom, any} | {:via, atom, any}
+  @type ip_address ::
+          atom
+          | charlist
+          | {:local, binary | charlist}
+          | {byte, byte, byte, byte}
+          | {char, char, char, char, char, char, char, char}
+
+  @type t :: %MLLP.Client{
+          socket: any(),
+          socket_address: String.t(),
+          address: ip_address(),
+          port: char(),
+          pending_reconnect: reference() | nil,
+          pid: pid() | nil,
+          telemetry_module: module() | nil,
+          tcp: module() | nil,
+          tls_opts: Keyword.t(),
+          socket_opts: Keyword.t()
+        }
+
+  defstruct socket: nil,
+            socket_address: "127.0.0.1:0",
+            address: {127, 0, 0, 1},
+            port: 0,
+            pending_reconnect: nil,
+            pid: nil,
+            telemetry_module: nil,
+            tcp: nil,
+            connect_failure: nil,
+            host_string: nil,
+            send_opts: %{},
+            tls_opts: [],
+            socket_opts: []
+
+  alias __MODULE__, as: State
+
+  ## API
+  @spec start_link(
+          address ::
+            :inet.ip4_address()
+            | :inet.ip6_address()
+            | String.t(),
+          port :: non_neg_integer(),
+          options :: [keyword()]
+        ) :: {:ok, pid()}
+
+  def start_link(address, port, options \\ []) do
+    checked_address = normalize_address!(address)
+    GenServer.start_link(__MODULE__, [address: checked_address, port: port] ++ options)
+  end
+
+  defp normalize_address!({_, _, _, _} = addr), do: addr
+  defp normalize_address!({_, _, _, _, _, _, _, _} = addr), do: addr
+
+  defp normalize_address!(addr) when is_binary(addr) do
+    case String.contains?(addr, ".") do
+      true ->
+        addr
+        |> String.to_charlist()
+        |> parse_address!()
+
+      false ->
+        # hostname
+        String.to_charlist(addr)
+    end
+  end
+
+  defp normalize_address!(addr) when is_list(addr), do: parse_address!(addr)
+
+  defp normalize_address!(addr) when is_atom(addr), do: addr
+
+  defp normalize_address!(addr), do: raise("Invalid server ip address : #{inspect(addr)}")
```

Comment:
Should we raise `MLLP.Client.Error` here?

Reply:
I think a  more appropriate error would be ArgumentError, wdyt?